### PR TITLE
*: Move stale read context to `TxnManager` 

### LINKED ISSUE #31954: Got error when stale read with scalar subquery
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t1(id int primary key, v int);
select * from t1 as of timestamp now() where id=(select id from t1 as of timestamp now() where id=1); --error
select (select * from t1 as of timestamp now() where id =1); --error
```

### 2. What did you expect to see? (Required)

no error

### 3. What did you see instead (Required)

(8024, 'invalid transaction')

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

master

```
mysql> select tidb_version();
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: None
Edition: Community
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: go1.16.3
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
But I think it happens in all versions

### LINKED ISSUE #33014: Add `GetReadTS` and `GetForUpdateTS` for `TxnManager`

### LINKED ISSUE #33832: Move the stale read context to the txn provider

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #33832, close #33014, close #31954

### What is changed and how it works?

- Provide `StalenessTxnContextProvider` to provide stale read context, including infoschema and readTS
- Adding two methods to `TxnManager` : `GetReadTS` and `GetForUpdateTS`
- After putting stale read informations to the context, there is no necessary to pass them to executor builder, so remove `snapshotTS` and `staleness` when constructing executerBuilder.
- After putting stale read informations to the context, the issue #31954 will be fixed.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

File: executor/executor_test.go [hunk 1]
```diff
@@ -3465,7 +3465,7 @@ func TestUnreasonablyClose(t *testing.T) {
 		&plannercore.PhysicalShuffle{},
 		&plannercore.PhysicalUnionAll{},
 	}
-	executorBuilder := executor.NewMockExecutorBuilderForTest(tk.Session(), is, nil, math.MaxUint64, false, "global")
+	executorBuilder := executor.NewMockExecutorBuilderForTest(tk.Session(), is, nil, "global")
```
Comment (JmPotato) [inline executor/executor_test.go:3468]: We should use `oracle.GlobalTxnScope` rather than `global`.

File: executor/executor.go [hunk 2]
```diff
@@ -1263,7 +1263,7 @@ func init() {
 			ctx = opentracing.ContextWithSpan(ctx, span1)
 		}
 
-		e := &executorBuilder{is: is, ctx: sctx}
+		e := newExecutorBuilder(sctx, is, nil, "")
```
Comment (JmPotato) [inline executor/executor.go:1266]: Why using an empty scope here?

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: executor/executor.go (hunk 2 above)
Comment (lcwangchao) [reply executor/executor.go:1266]: To keep the same configuration with the old code. But I think it's also fine to change it to `oracle.GlobalTxnScope`

**Review by JmPotato — APPROVED** [review]

**Pull request merged**